### PR TITLE
Refactoring CUDNN Find

### PR DESCRIPTION
**Pull request by borisfom**
Here, it does not break things that were working before - however, it does not yet fixes DoubleTensor even though similar changes Natalia did in her workspace, seemed to fix it.

### CODE REVIEW

File: test/test.lua [hunk 1]
```diff
@@ -1507,10 +1515,10 @@ math.randomseed(os.time())
 mytester = torch.Tester()
 mytester:add(cudnntest)
 
-if torch.random(1,2) == 1 then
-   cudnn.benchmark = true -- run manual auto-tuner
---   cudnn.verbose = true
-end
+-- if torch.random(1,2) == 1 then
```
Comment (soumith) [inline test/test.lua]: uncomment code here.

**Comment by soumith** [comment]
looks like a very nice refactor. added a small comment in the tests. seems good to go after that.

File: algo.lua [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        end
+
+        if extraBufferSizeInBytes < bufSize[1] then
+           self.extraBuffer:resize(math.ceil(bufSize[1]/self.extraBuffer.elementSize()))
+        end
+        return algType[0]
+end
+
+function algo.prepareHash(self, input_slice, output_slice)
+   local function shape(x)
```
Comment (ngimel) [inline algo.lua]: Can be replaced by one-liner from volumetricConvolution
https://github.com/soumith/cudnn.torch/blob/master/VolumetricConvolution.lua#L112-L114

Comment (borisfom) [inline algo.lua]: Thanks @ngimel ! I will use that one liner from VolumetricConvolution and then refactor it, too :). Any ideas why I may not have seen DoubleTensor errors fixed as you did ?

File: test/test.lua (hunk 1 above)
Comment (borisfom) [inline test/test.lua]: Sure! However, one of the reasons I have uncommented the condition is that it's always false: random generator needs to be seeded probably?

File: algo.lua (hunk 2 above)
Comment (borisfom) [inline algo.lua]: Thanks @soumith! I was thinking, maybe I should merge algo.lua into init.lua so that the methods would be of cudnn class ?

Comment (soumith) [inline algo.lua]: init.lua seems to be balloning in size and will be harder to maintain. better to keep algo.lua as is, i like this abstraction.

File: algo.lua [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+autotunerCache[1] = {} -- forward
+autotunerCache[2] = {} -- backwardFilter
+autotunerCache[3] = {} -- backwardData
+
+local function setupAlgo(self, algo_t, perf_t, findAPI, getAPI, wsAPI, algSearchMode, params)
+
+        local algType = ffi.new(algo_t, 1)
+
+        if cudnn.benchmark or cudnn.fastest then -- the manual auto-tuner is run
+            if autotunerCache[1][self.autotunerHash] then
```
Comment (ngimel) [inline algo.lua]: So if you've already called setupAlgo for forward for a particular convolution, and are now calling it for dgrad or wgrad,  autotunerCache[**1**][self.autotunerHash] will be non-nil, and you will return best forward algo for dgrad and wgrad? That won't work. If I'm not missing anything here, can you please fix it?

Comment (borisfom) [inline algo.lua]: You are absolutely right, I am surprised it even worked! I will use findAPI as the key instead of numbers.

Comment (borisfom) [inline algo.lua]: Please check now.

Comment (ngimel) [inline algo.lua]: Looks good in fb1bec1

File: VolumetricConvolution.lua [hunk 4]
```diff
@@ -2,95 +2,56 @@ local VolumetricConvolution, parent
...
 local errcheck = cudnn.errcheck
+local algo = require 'cudnn.algo'
 
-local autotunerCache = {}
-autotunerCache[1] = {} -- forward
-autotunerCache[2] = {} -- backwardFilter
-autotunerCache[3] = {} -- backwardData
+local Convolution = cudnn.SpatialConvolution
 
+function VolumetricConvolution:__init(nInputPlane, nOutputPlane,
```
Comment (anibali) [inline VolumetricConvolution.lua]: `cudnn.VolumetricConvolution` does not work for me after this PR was merged, and this looks kind of broken. Where are `dT` and `padT` in the arg list? Where is the call to the parent's __init?

Was VolumetricConvolution code tested before the merge?

File: VolumetricConvolution.lua [hunk 5]
```diff
@@ -291,119 +78,28 @@ function VolumetricConvolution:createIODescriptors(input)
...
-            self.convDesc[0], self.fwdAlgType[0],
-            self.extraBuffer:data(), self.extraBufferSizeInBytes,
-            zero:data(),
-            self.oDesc[0], self.output:data());
-   errcheck('cudnnAddTensor', cudnn.getHandle(),
-            one:data(),
-            self.biasDesc[0], self.bias:data(), one:data(),
-            self.oDescBias[0], self.output:data());
-   return self.output
+   return Convolution:updateOutput(input)
```
Comment (anibali) [inline VolumetricConvolution.lua]: There are multiple instances where `Convolution:funcname(...)` is used, which I believe should be `Convolution.funcname(self, ...)`

File: VolumetricConvolution.lua [hunk 6]
```diff
@@ -106,181 +67,7 @@ function VolumetricConvolution:createIODescriptors(input)
...
-        maxBufSize = math.max(maxBufSize, bufSize[1])
-
-        self.extraBuffer = self.extraBuffer or cudnn.getSharedWorkspace()
-        self.extraBufferSizeInBytes = self.extraBuffer:nElement() * 4 -- float
-        if maxBufSize > self.extraBufferSizeInBytes then
-            self.extraBuffer:resize(math.ceil(maxBufSize/4))
-            self.extraBufferSizeInBytes = maxBufSize
-        end
-        -----------------------------------------------------------------------
+         algo.prepareHash(self, input, output)
```
Comment (anibali) [inline VolumetricConvolution.lua]: Is this meant to refer to `self.output` instead of `output`?

**Comment by anibali** [comment]
The more I look, the more problems I am finding. I think it might be a good idea to revert VolumetricConvolution to a working state before attempting to fix these issues.

**Comment by borisfom** [comment]
Yes I did run the test.lua before the merge. I am actually about to make a new PR with remainder of the refactoring. And yes, your remarks about __init and ":" calls are addressed.

**Comment by anibali** [comment]
Perhaps test.lua does not cover VolumetricConvolution particularly well? It's probably a good idea to just run a simple forward pass as a sanity check:

```lua
cudnn.VolumetricConvolution(3, 10, 3,3,3, 1,1,1, 1,1,1)
  :forward(torch.CudaTensor(8, 5, 3, 100, 100))
```

File: TemporalConvolution.lua [hunk 7]
```diff
@@ -14,7 +16,8 @@ function TemporalConvolution:__init(inputFrameSize, outputFrameSize,
     local nOutputPlane = outputFrameSize
     self.inputFrameSize = inputFrameSize
     self.outputFrameSize = outputFrameSize
-    cudnn.SpatialConvolution.__init(self, nInputPlane, nOutputPlane, kW, kH, 1, dH,0,padH)
+    self.nDim = 4
```
Comment (Multihuntr) [inline TemporalConvolution.lua]: What's the 4th dimension here? I can only think of 3 dimensions that TemporalConvolution could have: {minibatch, number of features, sequence}
If VolumetricConvolution has nDim == 5, and SpatialConvolution has nDim == 4, then you would expect this to be nDim == 3, anyway.

**Comment by soumith** [comment]
Really sorry about this. I'm reverting that pr now. This was poor on our part. I'll improve the test coverage tomorrow

**Comment by borisfom** [comment]
Thanks Soumith - I was about to revert it myself. Next week I will do a new PR.

**Comment by soumith** [comment]
@anibali extremely sorry about this. Spent some time looking into it now. Looks like the tests weren't run before sending the PR (or were ignored). I revoked commit rights for the other maintainers. I am going to be manually testing and merging things (again) going forward.

**Comment by anibali** [comment]
No worries, these things happen sometimes. I'm immensely grateful for Torch and the work you (and others) put into it :smiley:

**Pull request merged**